File: impl/python/agentnet_py/agentmail.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional, Tuple

from .cbor import CborMap, CborValue, decode_canonical, encode_canonical
from .crypto import sha256, verify_ed25519_hash
from .sign import sign_ed25519_hash
from .markdown import validate_markdown_profile

AGENTMAIL_SIG_KEY = 14
AGENTMAIL_VERSION = 1


class AgentMailError(ValueError):
    pass


@dataclass
class AgentMailAttachment:
    content_hash: bytes
    size_bytes: int
    mime: str
    retrieval: Optional[List[str]]


@dataclass
class AgentMailMessagePayload:
    version: int
    message_id: str
    sender: str
    recipients: List[str]
    thread_id: Optional[str]
    reply_to: Optional[str]
    subject: Optional[str]
    markdown: str
    attachments: Optional[List[AgentMailAttachment]]
    intent_hashes: Optional[List[bytes]]
    receipt_hashes: Optional[List[bytes]]
    metadata: Optional[CborValue]
    ts: int
    expires: Optional[int]


@dataclass
class AgentMailMessage:
    payload: AgentMailMessagePayload
    signature: bytes


def parse_agentmail_payload(value: CborValue) -> AgentMailMessagePayload:
    entries = _expect_map(value)
    payload = AgentMailMessagePayload(
        version=_expect_u8(_get_required(entries, 0)),
        message_id=_expect_text(_get_required(entries, 1)),
        sender=_expect_text(_get_required(entries, 2)),
        recipients=_expect_text_array(_get_required(entries, 3)),
        thread_id=_optional_text(entries, 4),
        reply_to=_optional_text(entries, 5),
        subject=_optional_text(entries, 6),
        markdown=_expect_text(_get_required(entries, 7)),
        attachments=_optional_attachments(entries, 8),
        intent_hashes=_optional_hashes(entries, 9),
        receipt_hashes=_optional_hashes(entries, 10),
        metadata=_get_optional(entries, 11),
        ts=_expect_u64(_get_required(entries, 12)),
        expires=_optional_u64(entries, 13),
    )
    _validate_payload(payload)
    return payload


def parse_agentmail_message(value: CborValue) -> AgentMailMessage:
    payload, signature = _split_signed_map(value, AGENTMAIL_SIG_KEY)
    payload_obj = parse_agentmail_payload(payload)
    return AgentMailMessage(payload=payload_obj, signature=signature)


def decode_agentmail_message(data: bytes) -> AgentMailMessage:
    return parse_agentmail_message(decode_canonical(data))


def build_agentmail_message(payload: AgentMailMessagePayload, secret_key: bytes) -> bytes:
    _validate_payload(payload)
    payload_cbor = _agentmail_payload_to_cbor(payload)
    payload_bytes = encode_canonical(payload_cbor)
    digest = sha256(payload_bytes)
    signature = sign_ed25519_hash(secret_key, digest)
    full = _with_signature(payload_cbor, AGENTMAIL_SIG_KEY, signature)
    return encode_canonical(full)


def verify_agentmail_message(data: bytes, public_key: bytes) -> AgentMailMessagePayload:
    value = decode_canonical(data)
    payload, signature = _split_signed_map(value, AGENTMAIL_SIG_KEY)
    payload_bytes = encode_canonical(payload)
    digest = sha256(payload_bytes)
    verify_ed25519_hash(public_key, digest, signature)
    return parse_agentmail_payload(payload)


def _validate_payload(payload: AgentMailMessagePayload) -> None:
    if payload.version != AGENTMAIL_VERSION:
        raise AgentMailError("unsupported agentmail version")
    _ensure_nonempty(payload.message_id, "message_id required")
    _ensure_nonempty(payload.sender, "sender required")
    if not payload.recipients:
        raise AgentMailError("recipients required")
    for recipient in payload.recipients:
        _ensure_nonempty(recipient, "recipient required")
    if payload.thread_id is not None:
        _ensure_nonempty(payload.thread_id, "thread_id required")
    if payload.reply_to is not None:
        _ensure_nonempty(payload.reply_to, "reply_to required")
    if payload.subject is not None:
        _ensure_nonempty(payload.subject, "subject required")
    _ensure_nonempty(payload.markdown, "markdown required")
    validate_markdown_profile(payload.markdown)
    if payload.ts == 0:
        raise AgentMailError("timestamp required")
    if payload.expires is not None and payload.expires < payload.ts:
        raise AgentMailError("expires before timestamp")
    if payload.attachments is not None:
        if not payload.attachments:
            raise AgentMailError("attachments empty")
        for attachment in payload.attachments:
            _validate_attachment(attachment)
    if payload.intent_hashes is not None:
        if not payload.intent_hashes:
            raise AgentMailError("intent hashes empty")
        for digest in payload.intent_hashes:
            if len(digest) != 32:
                raise AgentMailError("intent hash length invalid")
    if payload.receipt_hashes is not None:
        if not payload.receipt_hashes:
            raise AgentMailError("receipt hashes empty")
        for digest in payload.receipt_hashes:
            if len(digest) != 32:
                raise AgentMailError("receipt hash length invalid")


def _validate_attachment(attachment: AgentMailAttachment) -> None:
    if len(attachment.content_hash) != 32:
        raise AgentMailError("attachment hash length invalid")
    if attachment.size_bytes <= 0:
        raise AgentMailError("attachment size required")
    _ensure_nonempty(attachment.mime, "attachment mime required")
    if attachment.retrieval is not None:
        if not attachment.retrieval:
            raise AgentMailError("attachment retrieval empty")
        for addr in attachment.retrieval:
            _ensure_nonempty(addr, "attachment retrieval invalid")


def _agentmail_payload_to_cbor(payload: AgentMailMessagePayload) -> CborValue:
    _validate_payload(payload)
    entries: List[Tuple[CborValue, CborValue]] = []
    entries.append((0, payload.version))
    entries.append((1, payload.message_id))
    entries.append((2, payload.sender))
    entries.append((3, list(payload.recipients)))
    if payload.thread_id is not None:
        entries.append((4, payload.thread_id))
    if payload.reply_to is not None:
        entries.append((5, payload.reply_to))
    if payload.subject is not None:
        entries.append((6, payload.subject))
    entries.append((7, payload.markdown))
    if payload.attachments is not None:
        entries.append((8, [_attachment_to_cbor(att) for att in payload.attachments]))
    if payload.intent_hashes is not None:
        entries.append((9, [bytes(digest) for digest in payload.intent_hashes]))
    if payload.receipt_hashes is not None:
        entries.append((10, [bytes(digest) for digest in payload.receipt_hashes]))
    if payload.metadata is not None:
        entries.append((11, payload.metadata))
    entries.append((12, payload.ts))
    if payload.expires is not None:
        entries.append((13, payload.expires))
    return CborMap(entries)


def _attachment_to_cbor(attachment: AgentMailAttachment) -> CborValue:
    _validate_attachment(attachment)
    entries: List[Tuple[CborValue, CborValue]] = []
    entries.append((0, bytes(attachment.content_hash)))
    entries.append((1, int(attachment.size_bytes)))
    entries.append((2, attachment.mime))
    if attachment.retrieval is not None:
        entries.append((3, list(attachment.retrieval)))
    return CborMap(entries)


def _optional_text(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[str]:
    value = _get_optional(entries, key)
    if value is None:
        return None
    return _expect_text(value)


def _optional_u64(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[int]:
    value = _get_optional(entries, key)
    if value is None:
        return None
    return _expect_u64(value)


def _optional_attachments(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[List[AgentMailAttachment]]:
    value = _get_optional(entries, key)
    if value is None:
        return None
    if not isinstance(value, list):
        raise AgentMailError("expected attachment array")
    attachments: List[AgentMailAttachment] = []
    for item in value:
        attachments.append(_parse_attachment(item))
    if not attachments:
        raise AgentMailError("attachments empty")
    return attachments


def _parse_attachment(value: CborValue) -> AgentMailAttachment:
    entries = _expect_map(value)
    content_hash = _expect_bytes_len(_get_required(entries, 0), 32)
    size_bytes = _expect_u64(_get_required(entries, 1))
    mime = _expect_text(_get_required(entries, 2))
    retrieval = _optional_text_array(entries, 3)
    attachment = AgentMailAttachment(
        content_hash=content_hash,
        size_bytes=size_bytes,
        mime=mime,
        retrieval=retrieval,
    )
    _validate_attachment(attachment)
    return attachment


def _optional_hashes(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[List[bytes]]:
    value = _get_optional(entries, key)
    if value is None:
        return None
    if not isinstance(value, list):
        raise AgentMailError("expected hash array")
    hashes: List[bytes] = []
    for item in value:
        hashes.append(_expect_bytes_len(item, 32))
    if not hashes:
        raise AgentMailError("hash array empty")
    return hashes


def _optional_text_array(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[List[str]]:
    value = _get_optional(entries, key)
    if value is None:
        return None
    return _expect_text_array(value)


def _expect_map(value: CborValue) -> List[Tuple[CborValue, CborValue]]:
    if isinstance(value, CborMap):
        return list(value.entries)
    raise AgentMailError("expected cbor map")


def _get_required(entries: List[Tuple[CborValue, CborValue]], key: int) -> CborValue:
    for entry_key, entry_value in entries:
        if _is_uint(entry_key) and entry_key == key:
            return entry_value
    raise AgentMailError("missing required key")


def _get_optional(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[CborValue]:
    for entry_key, entry_value in entries:
        if _is_uint(entry_key) and entry_key == key:
            return entry_value
    return None


def _expect_u8(value: CborValue) -> int:
    if _is_uint(value) and 0 <= value <= 255:
        return int(value)
    raise AgentMailError("expected u8")


def _expect_u64(value: CborValue) -> int:
    if _is_uint(value):
        return int(value)
    raise AgentMailError("expected u64")


def _expect_text(value: CborValue) -> str:
    if isinstance(value, str):
        return value
    raise AgentMailError("expected text")


def _expect_text_array(value: CborValue) -> List[str]:
    if not isinstance(value, list):
        raise AgentMailError("expected array of text")
    out: List[str] = []
    for item in value:
        out.append(_expect_text(item))
    return out


def _expect_bytes(value: CborValue) -> bytes:
    if isinstance(value, (bytes, bytearray)):
        return bytes(value)
    raise AgentMailError("expected bytes")


def _expect_bytes_len(value: CborValue, length: int) -> bytes:
    data = _expect_bytes(value)
    if len(data) != length:
        raise AgentMailError("invalid length")
    return data


def _split_signed_map(value: CborValue, signature_key: int) -> Tuple[CborMap, bytes]:
    entries = _expect_map(value)
    payload_entries: List[Tuple[CborValue, CborValue]] = []
    signature: Optional[bytes] = None
    for key, val in entries:
        if _is_uint(key) and key == signature_key:
            signature = _expect_bytes(val)
        else:
            payload_entries.append((key, val))
    if signature is None:
        raise AgentMailError("signature missing")
    return CborMap(payload_entries), signature


def _with_signature(payload: CborValue, signature_key: int, signature: bytes) -> CborMap:
    entries = _expect_map(payload)
    entries.append((signature_key, bytes(signature)))
    return CborMap(entries)


def _ensure_nonempty(value: str, message: str) -> None:
    if value.strip() == "":
        raise AgentMailError(message)


def _is_uint(value: CborValue) -> bool:
    return isinstance(value, int) and value >= 0 and not isinstance(value, bool)

File: impl/rust/crates/anetsdk/src/agentmail.rs
use crate::markdown::validate_markdown_profile;
use crate::schema::{
    expect_bytes_len, expect_map, expect_text, expect_text_array, expect_u64, expect_u8,
    get_optional, get_required,
};
use crate::signed::{split_signed_map, with_signature};
use crate::{
    decode_canonical, encode_canonical, sha256, sign_ed25519_hash, verify_ed25519_hash, CborValue,
    Error,
};

const AGENTMAIL_SIG_KEY: u64 = 14;
const AGENTMAIL_VERSION: u8 = 1;

#[derive(Debug, Clone)]
pub struct AgentMailAttachment {
    pub content_hash: Vec<u8>,
    pub size_bytes: u64,
    pub mime: String,
    pub retrieval: Option<Vec<String>>,
}

#[derive(Debug, Clone)]
pub struct AgentMailMessagePayload {
    pub version: u8,
    pub message_id: String,
    pub sender: String,
    pub recipients: Vec<String>,
    pub thread_id: Option<String>,
    pub reply_to: Option<String>,
    pub subject: Option<String>,
    pub markdown: String,
    pub attachments: Option<Vec<AgentMailAttachment>>,
    pub intent_hashes: Option<Vec<Vec<u8>>>,
    pub receipt_hashes: Option<Vec<Vec<u8>>>,
    pub metadata: Option<CborValue>,
    pub ts: u64,
    pub expires: Option<u64>,
}

#[derive(Debug, Clone)]
pub struct AgentMailMessage {
    pub payload: AgentMailMessagePayload,
    pub signature: Vec<u8>,
}

pub fn parse_agentmail_payload(value: &CborValue) -> Result<AgentMailMessagePayload, Error> {
    let map = expect_map(value)?;
    let version = expect_u8(get_required(&map, 0)?)?;
    let message_id = expect_text(get_required(&map, 1)?)?;
    let sender = expect_text(get_required(&map, 2)?)?;
    let recipients = expect_text_array(get_required(&map, 3)?)?;
    let thread_id = match get_optional(&map, 4) {
        Some(value) => Some(expect_text(value)?),
        None => None,
    };
    let reply_to = match get_optional(&map, 5) {
        Some(value) => Some(expect_text(value)?),
        None => None,
    };
    let subject = match get_optional(&map, 6) {
        Some(value) => Some(expect_text(value)?),
        None => None,
    };
    let markdown = expect_text(get_required(&map, 7)?)?;
    let attachments = match get_optional(&map, 8) {
        Some(value) => Some(parse_attachments(value)?),
        None => None,
    };
    let intent_hashes = match get_optional(&map, 9) {
        Some(value) => Some(expect_hash_array(value)?),
        None => None,
    };
    let receipt_hashes = match get_optional(&map, 10) {
        Some(value) => Some(expect_hash_array(value)?),
        None => None,
    };
    let metadata = get_optional(&map, 11).cloned();
    let ts = expect_u64(get_required(&map, 12)?)?;
    let expires = match get_optional(&map, 13) {
        Some(value) => Some(expect_u64(value)?),
        None => None,
    };

    let payload = AgentMailMessagePayload {
        version,
        message_id,
        sender,
        recipients,
        thread_id,
        reply_to,
        subject,
        markdown,
        attachments,
        intent_hashes,
        receipt_hashes,
        metadata,
        ts,
        expires,
    };
    payload.validate()?;
    Ok(payload)
}

pub fn parse_agentmail_message(value: &CborValue) -> Result<AgentMailMessage, Error> {
    let (payload_entries, signature) = split_signed_map(value, AGENTMAIL_SIG_KEY)?;
    let payload_value = CborValue::Map(payload_entries);
    let payload = parse_agentmail_payload(&payload_value)?;
    Ok(AgentMailMessage { payload, signature })
}

pub fn decode_agentmail_message(data: &[u8]) -> Result<AgentMailMessage, Error> {
    let value = decode_canonical(data)?;
    parse_agentmail_message(&value)
}

pub fn build_agentmail_message(
    payload: &AgentMailMessagePayload,
    secret_key: &[u8],
) -> Result<Vec<u8>, Error> {
    payload.validate()?;
    let payload_value = payload.to_cbor()?;
    let payload_cbor = encode_canonical(&payload_value)?;
    let hash = sha256(&payload_cbor);
    let sig = sign_ed25519_hash(secret_key, &hash)?;
    let full = with_signature(&payload_value, AGENTMAIL_SIG_KEY, sig)?;
    encode_canonical(&full)
}

pub fn verify_agentmail_message(
    data: &[u8],
    public_key: &[u8],
) -> Result<AgentMailMessagePayload, Error> {
    let value = decode_canonical(data)?;
    let (payload_entries, signature) = split_signed_map(&value, AGENTMAIL_SIG_KEY)?;
    let payload_value = CborValue::Map(payload_entries);
    let payload_cbor = encode_canonical(&payload_value)?;
    let hash = sha256(&payload_cbor);
    verify_ed25519_hash(public_key, &hash, &signature)?;
    parse_agentmail_payload(&payload_value)
}

impl AgentMailAttachment {
    pub fn to_cbor(&self) -> Result<CborValue, Error> {
        if self.content_hash.len() != 32 {
            return Err(Error::Cbor("invalid attachment hash length"));
        }
        if self.size_bytes == 0 {
            return Err(Error::Cbor("attachment size required"));
        }
        if self.mime.trim().is_empty() {
            return Err(Error::Cbor("attachment mime required"));
        }
        let mut entries = Vec::new();
        entries.push((
            CborValue::Unsigned(0),
            CborValue::Bytes(self.content_hash.clone()),
        ));
        entries.push((CborValue::Unsigned(1), CborValue::Unsigned(self.size_bytes)));
        entries.push((CborValue::Unsigned(2), CborValue::Text(self.mime.clone())));
        if let Some(addrs) = &self.retrieval {
            if addrs.is_empty() {
                return Err(Error::Cbor("attachment retrieval empty"));
            }
            entries.push((
                CborValue::Unsigned(3),
                CborValue::Array(addrs.iter().map(|s| CborValue::Text(s.clone())).collect()),
            ));
        }
        Ok(CborValue::Map(entries))
    }
}

impl AgentMailMessagePayload {
    pub fn to_cbor(&self) -> Result<CborValue, Error> {
        self.validate()?;
        let mut entries = Vec::new();
        entries.push((
            CborValue::Unsigned(0),
            CborValue::Unsigned(self.version as u64),
        ));
        entries.push((
            CborValue::Unsigned(1),
            CborValue::Text(self.message_id.clone()),
        ));
        entries.push((CborValue::Unsigned(2), CborValue::Text(self.sender.clone())));
        entries.push((
            CborValue::Unsigned(3),
            CborValue::Array(
                self.recipients
                    .iter()
                    .map(|s| CborValue::Text(s.clone()))
                    .collect(),
            ),
        ));
        if let Some(thread_id) = &self.thread_id {
            entries.push((CborValue::Unsigned(4), CborValue::Text(thread_id.clone())));
        }
        if let Some(reply_to) = &self.reply_to {
            entries.push((CborValue::Unsigned(5), CborValue::Text(reply_to.clone())));
        }
        if let Some(subject) = &self.subject {
            entries.push((CborValue::Unsigned(6), CborValue::Text(subject.clone())));
        }
        entries.push((
            CborValue::Unsigned(7),
            CborValue::Text(self.markdown.clone()),
        ));
        if let Some(attachments) = &self.attachments {
            let mut items = Vec::with_capacity(attachments.len());
            for attachment in attachments {
                items.push(attachment.to_cbor()?);
            }
            entries.push((CborValue::Unsigned(8), CborValue::Array(items)));
        }
        if let Some(intent_hashes) = &self.intent_hashes {
            entries.push((
                CborValue::Unsigned(9),
                CborValue::Array(
                    intent_hashes
                        .iter()
                        .map(|h| CborValue::Bytes(h.clone()))
                        .collect(),
                ),
            ));
        }
        if let Some(receipt_hashes) = &self.receipt_hashes {
            entries.push((
                CborValue::Unsigned(10),
                CborValue::Array(
                    receipt_hashes
                        .iter()
                        .map(|h| CborValue::Bytes(h.clone()))
                        .collect(),
                ),
            ));
        }
        if let Some(metadata) = &self.metadata {
            entries.push((CborValue::Unsigned(11), metadata.clone()));
        }
        entries.push((CborValue::Unsigned(12), CborValue::Unsigned(self.ts)));
        if let Some(expires) = self.expires {
            entries.push((CborValue::Unsigned(13), CborValue::Unsigned(expires)));
        }
        Ok(CborValue::Map(entries))
    }

    pub fn validate(&self) -> Result<(), Error> {
        if self.version != AGENTMAIL_VERSION {
            return Err(Error::Cbor("unsupported agentmail version"));
        }
        if self.message_id.trim().is_empty() {
            return Err(Error::Cbor("message_id required"));
        }
        if self.sender.trim().is_empty() {
            return Err(Error::Cbor("sender required"));
        }
        if self.recipients.is_empty() {
            return Err(Error::Cbor("recipients required"));
        }
        for recipient in &self.recipients {
            if recipient.trim().is_empty() {
                return Err(Error::Cbor("recipient required"));
            }
        }
        if let Some(thread_id) = &self.thread_id {
            if thread_id.trim().is_empty() {
                return Err(Error::Cbor("thread_id required"));
            }
        }
        if let Some(reply_to) = &self.reply_to {
            if reply_to.trim().is_empty() {
                return Err(Error::Cbor("reply_to required"));
            }
        }
        if let Some(subject) = &self.subject {
            if subject.trim().is_empty() {
                return Err(Error::Cbor("subject required"));
            }
        }
        if self.markdown.trim().is_empty() {
            return Err(Error::Cbor("markdown required"));
        }
        validate_markdown_profile(&self.markdown)?;
        if self.ts == 0 {
            return Err(Error::Cbor("timestamp required"));
        }
        if let Some(expires) = self.expires {
            if expires < self.ts {
                return Err(Error::Cbor("expires before timestamp"));
            }
        }
        if let Some(attachments) = &self.attachments {
            if attachments.is_empty() {
                return Err(Error::Cbor("attachments empty"));
            }
            for attachment in attachments {
                if attachment.content_hash.len() != 32 {
                    return Err(Error::Cbor("attachment hash length invalid"));
                }
                if attachment.size_bytes == 0 {
                    return Err(Error::Cbor("attachment size required"));
                }
                if attachment.mime.trim().is_empty() {
                    return Err(Error::Cbor("attachment mime required"));
                }
                if let Some(addrs) = &attachment.retrieval {
                    if addrs.is_empty() {
                        return Err(Error::Cbor("attachment retrieval empty"));
                    }
                    for addr in addrs {
                        if addr.trim().is_empty() {
                            return Err(Error::Cbor("attachment retrieval invalid"));
                        }
                    }
                }
            }
        }
        if let Some(hashes) = &self.intent_hashes {
            if hashes.is_empty() {
                return Err(Error::Cbor("intent hashes empty"));
            }
            for hash in hashes {
                if hash.len() != 32 {
                    return Err(Error::Cbor("intent hash length invalid"));
                }
            }
        }
        if let Some(hashes) = &self.receipt_hashes {
            if hashes.is_empty() {
                return Err(Error::Cbor("receipt hashes empty"));
            }
            for hash in hashes {
                if hash.len() != 32 {
                    return Err(Error::Cbor("receipt hash length invalid"));
                }
            }
        }
        Ok(())
    }
}

fn parse_attachments(value: &CborValue) -> Result<Vec<AgentMailAttachment>, Error> {
    match value {
        CborValue::Array(items) => {
            let mut out = Vec::with_capacity(items.len());
            for item in items {
                out.push(parse_attachment(item)?);
            }
            Ok(out)
        }
        _ => Err(Error::Cbor("expected attachment array")),
    }
}

fn parse_attachment(value: &CborValue) -> Result<AgentMailAttachment, Error> {
    let map = expect_map(value)?;
    let content_hash = expect_bytes_len(get_required(&map, 0)?, 32)?;
    let size_bytes = expect_u64(get_required(&map, 1)?)?;
    let mime = expect_text(get_required(&map, 2)?)?;
    let retrieval = match get_optional(&map, 3) {
        Some(value) => Some(expect_text_array(value)?),
        None => None,
    };
    let attachment = AgentMailAttachment {
        content_hash,
        size_bytes,
        mime,
        retrieval,
    };
    if attachment.size_bytes == 0 {
        return Err(Error::Cbor("attachment size required"));
    }
    if attachment.mime.trim().is_empty() {
        return Err(Error::Cbor("attachment mime required"));
    }
    if let Some(addrs) = &attachment.retrieval {
        if addrs.is_empty() {
            return Err(Error::Cbor("attachment retrieval empty"));
        }
        for addr in addrs {
            if addr.trim().is_empty() {
                return Err(Error::Cbor("attachment retrieval invalid"));
            }
        }
    }
    Ok(attachment)
}

fn expect_hash_array(value: &CborValue) -> Result<Vec<Vec<u8>>, Error> {
    match value {
        CborValue::Array(items) => {
            let mut out = Vec::with_capacity(items.len());
            for item in items {
                out.push(expect_bytes_len(item, 32)?);
            }
            Ok(out)
        }
        _ => Err(Error::Cbor("expected hash array")),
    }
}
